package dfu;

message command {
	enum op_code_t {
		RESET = 0;
		INIT = 1;
	}

	optional op_code_t op_code = 1;
	optional init init_packet = 2;
}

message init {
	optional uint32 fw_version = 1 [default = 4294967295]; // 0xFFFFFFFF
	optional uint32 hw_version = 2 [default = 4294967295]; // 0xFFFFFFFF
	repeated uint32 sd_req = 3 [packed = true];

	enum fw_type_t {
		SOFTDEVICE = 1;
		BOOTLOADER = 2;
		SOFTDEVICE_BOOTLOADER = 3;
		APPLICATION = 4;
	}
	optional fw_type_t type = 4;

	message sd_bl_size_t {
		required uint32 sd_size = 1;
  		required uint32 bl_size = 2;
	}
	oneof sizes {
		uint32 sd_size = 11;
		uint32 bl_size = 12;
		sd_bl_size_t sd_bl_size = 13;
		uint32 app_size = 14;
	}

	enum hash_type_t {
		NO_HASH = 0;
		CRC = 1;
		SHA128 = 2;
	}
	optional hash_type_t hash_type = 5;
	optional bytes hash = 6;

	enum signature_type_t {
		NO_SIGNATURE = 0;
		P256 = 1;
		ED25519 = 2;
	}
	optional signature_type_t signature_type = 7;
	optional bytes signature = 8;

	extensions 9 to 10;
}